[NCCL] Support NCCL Send/Recv

mingzhe09088 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45873 [NCCL] support send/recv to/from self when communicator is created on demand
* #44922 [NCCL] create NCCL communicator for send/recv on demand
* **#44921 [NCCL] Support NCCL Send/Recv**

This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.


Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

Commit:
[WIP][NCCL] Support NCCL Send/Recv

This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.

Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

Commit:
Update on "[WIP][NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.

Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +784,61 @@ def recv(tensor,
+    Arguments:
+        op_list: list of point-to-point operations(type of each operations is either
+        ``torch.distributed.isend`` or ``torch.distributed.irecv``.
+        tensor_list (list[Tensor]): list of send or recv tensors.
+        peer_list (list[int]): list of peer ranks to send to or receive from.
+        group_list (list[ProcessGroup], Optional): list of groups to operator on.
+        tag_list (list[int], Optional): list of tags to match send with recv.
```
The docs should mention that each of these lists should be the same length as the `op_list`. Also, for each argument let's mention in general what index `i` of that list refers to wrt to the ith operation in `op_list`.

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +784,61 @@ def recv(tensor,
+
+    Return:
+        a list of distributed request objects returned by calling the corresponding
+        op in the op_list.
```
Can we add an example to the doc as well?

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +784,61 @@ def recv(tensor,
+        group_list (list[ProcessGroup], Optional): list of groups to operator on.
+        tag_list (list[int], Optional): list of tags to match send with recv.
+
+    Return:
```
nit: Return**s**

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +784,61 @@ def recv(tensor,
+        ``torch.distributed.isend`` or ``torch.distributed.irecv``.
+        tensor_list (list[Tensor]): list of send or recv tensors.
+        peer_list (list[int]): list of peer ranks to send to or receive from.
+        group_list (list[ProcessGroup], Optional): list of groups to operator on.
```
What is the use case of having a `group_list` here instead of just having a single group? For ProcessGroupNCCL, as per NCCL's guidance we can't use multiple process groups anyways.

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +784,61 @@ def recv(tensor,
+    _check_op_list(op_list, "op_list")
+    _check_input_length(op_list, tensor_list, peer_list, group_list, tag_list)
+
+    backend = get_backend(group.WORLD if group_list is None else group_list[0])
```
If we are planning to support a `group_list`, we should ensure that every group in the list has the same backend and also add this requirement to our docs.

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +784,61 @@ def recv(tensor,
+        tag_list (list[int], Optional): list of tags to match send with recv.
+
+    Return:
+        a list of distributed request objects returned by calling the corresponding
```
nit: start with capital **A**

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -756,7 +757,26 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
   streamVal.reserve(devices.size());
 
-  // Create the NCCL communicators for each GPU
+  // [Group Start/End Note] This is used to ensure that nccl communicator will be created
+  // before communication primitives are called. Let's look at this example:
+  // Using the batch_isend_irecv to send a tensor to a target process. On the sender side,
+  // the corresponding underlying NCCL calls will look like
+  //   ncclGroupStart() // This is in batch_isend_irecv
+  //   ncclGroupStart() // This is [Note 1]
+  //   ncclCommInitRank() // Inside NCCLCOmm::create
+  //   ncclSend()
+  //   ncclGroupEnd() // This is [Note 2]
+  //   ncclGroupEnd() // This is in batch_isend_irecv
+  // With this pattern, the nccl communicator will be created in the last ncclGroupEnd
+  // which means when ncclSend is processed, the passed communicator argument is NULL which will
+  // lead to runtime error. So we need to "close" all active nccl groups to ensure
+  // nccl communicator is actually created before encountering any communication calls.
+  // This is why we need the following for loop.
+  for (size_t i = 0; i < ncclActiveGroupCounter_; ++i) {
+    C10D_NCCL_CHECK(ncclGroupEnd());
+  }
```
Instead of doing this, can we do this (which might be simpler):

1. In `_group_manager` just increment/decrement `ncclActiveGroupCounter_` via groupStart/groupEnd, but don't actually call `ncclGroupEnd` or `ncclGroupStart`.
2. In `pointTopoint`, we add the appropriate number of `ncclGroupStart` and `ncclGroupEnd` calls around the appropriate kernel?

This will probably work for our current use case, but not sure if we plan to use groupStart and groupEnd as a general API in the future. If so, this probably won't work since groupStart and groupEnd might be used with collectives and other ops too.

mingzhe09088 replied:
ncclGroupEnd needs to be called by the last Op in the op_list, we will need to maintain another counter to indicate end of the list. So existing design with one counter might be easier.

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -713,6 +713,12 @@ They are used in specifying strategies for reduction collectives, e.g.,
       .def(py::init<>())
       .def_readwrite("is_high_priority", &::c10d::ProcessGroupNCCL::Options::isHighPriorityStream)
       .def_readwrite("op_timeout", &::c10d::ProcessGroupNCCL::Options::opTimeout);
+  processGroupNCCL.def_static("group_start", []() {
```
Probably `_group_start` and `_group_end` to indicate its private?

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1006,6 +1032,72 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
+
+  post(ncclStreams_[key]);
+
+  // Event should only be recorded after the ncclGroupEnd()
```
There is no `ncclGroupEnd` here?

mingzhe09088 replied:
Why do we need ncclGroupEnd here?

pritamdamania87 replied:
I'm referring to the comment "// Event should only be recorded after the ncclGroupEnd()", this probably doesn't apply here right?

mingzhe09088 replied:
Oh, that's referring to the ncclGroupEnd inside the nccl_group_guard. We are fine here.

pritamdamania87 replied:
I see maybe expand on the comment a bit to indicate its referring to `nccl_group_guard`.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1327,24 +1506,10 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::scatter(
     std::vector<at::Tensor>& /* unused */,
     int /* unused */) {
-  throw std::runtime_error("ProcessGroupNCCL does not support recv");
+  throw std::runtime_error("ProcessGroupNCCL does not support recv Anysource");
```
nit `recvAnysource`

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -571,6 +574,56 @@ def test_backend_group(self):
+            self._barrier()
+
+        @skip_if_no_gpu
+        @unittest.skipIf(BACKEND != "nccl", "NCCL Send Recv Only")
+        def test_send_recv_nccl(self):
```
Can we test error scenarios for `batch_isend_irecv`? Some examples:

1. Invalid ops.
2. Invalid peer_list, group_list, tag_list etc. lengths.
3. Invalid mix of nccl and gloo backends.
4. Invalid set of tensors (ex: CPU tensors, sparse tensors etc.)

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -571,6 +574,56 @@ def test_backend_group(self):
 
+        # NCCL Batch SEND RECV
+        @skip_if_no_gpu
+        @unittest.skipIf(BACKEND != "nccl", "NCCL Send Recv Only")
```
Can we add a test with multiple tags?

mingzhe09088 replied:
NCCL doesn't use tags. Do we still want to add that test?

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +784,61 @@ def recv(tensor,
+    in each passed parameters, etc.
+
+    Arguments:
+        op_list: list of point-to-point operations(type of each operations is either
```
Should we also mention here that the order of the send/recv in the list matters and that is how on the remote end the corresponding send/recv are matched up?

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -571,6 +574,56 @@ def test_backend_group(self):
 
+        # NCCL Batch SEND RECV
+        @skip_if_no_gpu
+        @unittest.skipIf(BACKEND != "nccl", "NCCL Send Recv Only")
```
Doesn't this work for Gloo as well?

mingzhe09088 replied:
Based on https://pytorch.org/docs/stable/distributed.html, seems that we haven't officially claimed GLOO supports send/recv on GPUs. So I just tested NCCL send/recv here.

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -571,6 +574,56 @@ def test_backend_group(self):
+            self._barrier()
+
         # SEND RECV
         @unittest.skipIf(BACKEND == "nccl", "Nccl does not support send/recv")
```
Why can't we just reuse this test from gloo insted of duplicating it above?

mingzhe09088 replied:
This is a CPU test so we can't test NCCL here.

jiayisuse commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +784,61 @@ def recv(tensor,
 
 
+@contextlib.contextmanager
+def _group_manager(backend):
```
nit: _group_manager could be confused with process group. maybe _batch_p2p_manager ?

jiayisuse commented on torch/distributed/distributed_c10d.py:
```diff
@@ -245,6 +246,36 @@ def _check_tensor_list(param, param_name):
+    expected_length = len(op_list)
+    check_list = [tensor_list, peer_list]
+    if group_list is not None:
+        check_list.append(group_list)
```
Didn't quite get why we enforce group_list to be the same length with op_list. I guess a common use case is to have a local process group and issue batched isend/irecv through this single process group? And multiple NCCL communicators (process groups) of a process is not safe according to Nvidia

mingzhe09088 replied:
I want to keep the interface same as isend/irecv to provide flexibility for users. With the current way, it can cover cases 1)all the isend/irecv are in a single process group, users can create a list and pass it to group_list. 2) different ops are from different groups. I agree it is not safe for NCCL but this batch interface could potentially be used for other backends and not just GPUs.

jiayisuse commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +784,61 @@ def recv(tensor,
+            ProcessGroupNCCL.group_end()
+
+
+def batch_isend_irecv(op_list,
+                      tensor_list,
+                      peer_list,
+                      group_list=None,
+                      tag_list=None):
```
Feeling this may not be the most intrinsic way. Users should firstly have a series of p2p calls in mind. To use this batched interface, they have to split these p2p calls into op list, tensor list and peer list. Could we have a class object of for p2pOp, recording isend or irecv, peer, tensor, tag, and pass into this interface as a list of [p2pOp]

mingzhe09088 replied:
Not sure if I get the new design?

pritamdamania87 replied:
I like @jiayisuse's suggestion here and I think the design you outline above is a good approach.

mingzhe09088 replied:
Let me add p2pOp.

mingzhe09088 replied:
Let me add p2pOP.

jiayisuse commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -479,6 +483,22 @@ class ProcessGroupNCCL : public ProcessGroup {
       PostProcess post);
 
+  // Helper that encapsulates work shared across point-to-point communication
+  // primitives.It is the same structure as the helper used for collective
```
nit: blank space before `It is the same structure`

jiayisuse commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -198,10 +198,13 @@ def _lock():
 
 
-def _build_tensor(size, value=None, dtype=torch.float):
+def _build_tensor(size, value=None, dtype=torch.float, rank=None, cuda=False, rank_to_GPU=None):
```
Maybe just provide torch.device rather than rank + cuda + rank_to_GPU?

jiayisuse commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -756,7 +757,26 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
+  // the corresponding underlying NCCL calls will look like
+  //   ncclGroupStart() // This is in batch_isend_irecv
+  //   ncclGroupStart() // This is [Note 1]
+  //   ncclCommInitRank() // Inside NCCLCOmm::create
```
nit: NCCLComm::create

Commit:
Update on "[WIP][NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.

Todo: add more tests. 

Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

Commit:
Update on "[NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.


Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

Commit:
Update on "[NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.


Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

Commit:
Update on "[NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.


Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +778,82 @@ def recv(tensor,
+            ProcessGroupNCCL._group_end()
+
+
+def batch_isend_irecv(op_list,
```
We should clarify the behavior of this API for Gloo. As per my understanding we should probably mention in the docs that this API works for CPU tensors on Gloo and as a result we should also add tests validating this works correctly for Gloo.

Commit:
Update on "[NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.


Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

Commit:
Update on "[NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.


Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -778,6 +766,34 @@ def recv(tensor,
 
+class P2POp(object):
+    """
+    A class to save point-to-point operations for ``batch_isend_irecv``.
```
"A class to **build** point-to-point operations..." sounds better.

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -778,6 +766,34 @@ def recv(tensor,
+    """
+    A class to save point-to-point operations for ``batch_isend_irecv``.
+
+    This calss saves the type of P2P operation, communication buffer, peer rank,
```
nit: s/calss/class, build instead of save.

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -778,6 +766,34 @@ def recv(tensor,
+    ``batch_isend_irecv`` for point-to-point communications.
+
+    Arguments:
+        op (callable): A function to send data to or receive data from a peer process.
```
Mention in the docs that this should be `dist.isend` or `dist.irecv`.

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -789,30 +805,18 @@ def _batch_p2p_manager(backend):
-        groups in ``group_list`` should use the same backend.
-        tag_list (list[int], Optional): list of tags to match send with recv.
+        p2p_op_list: A list of point-to-point operations(type of each operator is
+        ``torch.distributed.P2POp``. The order of the isend/irecv in the list
```
nit: close the brackets here.

pritamdamania87 reviewed (approved):
LGTM, @jiayisuse can you take another look at the PR too?

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +766,92 @@ def recv(tensor,
+
+def batch_isend_irecv(p2p_op_list):
+    """
+    Send or Receive tensors asynchronously and return a list of requests.
```
Mention "batch" here. Ex: "Send or Receive a batch of tensors asynchronously"

Commit:
Update on "[NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.


Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

Commit:
Update on "[NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.


Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

Commit:
Update on "[NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.


Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +766,94 @@ def recv(tensor,
         return src
 
 
+class P2POp(object):
```
Since we don't have test coverage for OSS on this, shall we mark it as a private API for now (start with an underscore) and make sure its doc does not show in our API page?

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -753,6 +766,94 @@ def recv(tensor,
+            ProcessGroupNCCL._group_end()
+
+
+def batch_isend_irecv(p2p_op_list):
```
ditto

Commit:
Update on "[NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.


Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

Commit:
Update on "[NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.


Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]

Commit:
Update on "[NCCL] Support NCCL Send/Recv"


This diff adds support for Process Group point-to-point operations on NCCL backend based on ncclSend/ncclRecv. See https://github.com/pytorch/pytorch/issues/43995 for more context.


Differential Revision: [D23709848](https://our.internmc.facebook.com/intern/diff/D23709848/)

[ghstack-poisoned]
